check.js

```diff
@@ -0,0 +1,459 @@
+    res.reason = reasons.spawnPrefix;
+    return res;
+  }
+
+  res.result = true
+  return res;
+}
+
+/**
+ * Check if an address can set a spawn proxy for the given ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canSetSpawnProxy(contracts, prefix, address) {
+  let res = { result: false };
+  let parentShip = await ships.getShip(contracts, prefix);
+  // must be the owner of the ship
+  if (!await ships.isOwner(contracts, parentShip, address)) {
+    res.reason = reasons.permission;
+    return res;
+  }
+  // the ship must be active
+  if (!await ships.isActive(contracts, parentShip)) {
+    res.reason = reasons.inactive;
+    return res;
+  }
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the sender address can send the provided ship to the target
+ * address.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} sender - Sender's address.
+ * @param {String} target - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canTransferShip(contracts, ship, source, target) {
+  let res     = { result: false };
+  // prevent burning of ships
+  if (utils.isZeroAddress(target))
+  {
+    res.reason = reasons.zero;
+    return res;
+  }
+
+  let shipObj = await ships.getShip(contracts, ship);
+  // must be either the owner of the ship,
+  // a transfer proxy for the ship,
+  // or an operator for the owner
+  if (!ships.isOwner(contracts, shipObj, source) &&
+      !ships.isTransferProxy(contracts, shipObj, source) &&
+      !await ships.isOperator(contracts, shipObj.owner, source))
+  {
+    res.reason = reasons.permission;
+    return res;
+  }
+
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the address can set a transfer proxy for the ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canSetTransferProxy(contracts, ship, address) {
+  let res     = { result: false };
+  let shipObj = await ships.getShip(contracts, ship);
+  // must be either the owner of the ship,
+  // or an operator for the owner
+  if (!ships.isOwner(contracts, shipObj, address) &&
+      !await ships.isOperator(contracts, shipObj.owner, address))
+  {
+    res.reason = reasons.permission;
+    return res;
+  }
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the address can configure public keys for the ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canConfigureKeys(contracts, ship, address) {
```

Comment:
Any reason you're not just using `checkActiveShipManager` here, as you do in other places, and as #1 does?

Reply:
Huh, nope.  I think I was on the track of thinking that `checkActiveShipManager` should be replaced by `canConfigureKeys`, or vice versa, or merged together as something else (since they check for the same thing), but forgot to follow up on it.  Will fix.

Both boil down to `canManageAndShipIsActive`, and so in theory could be replaced by that. 
 But that's a mouthful. 😄

Reply:
Woah, no, that's the opposite of what I wanted here lol.

`canManageAndShipIsActive` is a specific check about the state it describes. Other checks should use this as a building block for their often more complex checks.  
`canConfigureKeys` is a function check about the ability to call that specific function. It just so happens that the only check that does is `canManageAndShipIsActive`, and that's totally fine.

Surely that makes more sense than having function checks depend on function checks? The above also more closely matches to the contracts themselves, something I should double-check wasn't lost-in-translation for other checks either.

Reply:
So, to be clear on my vision here:

- There should be a `canX` check for every function the Constitution contract exposes,
- There should be helper functions for parts of those checks in the same way the Constitution contract has them.

That way we stay nice and close to the contracts, and don't force the user to reason about what it takes to be able to successfully call a specific function. Sensical y/n?

Reply:
Ahhh right, gotcha.  I missed that you were matching those helper functions with the actual assertions in the contracts -- I thought it was a redundant internal thing that I could safely eliminate.  Ok, I'll put `activeShipManager` back in.

I think this is the only one that actually got *changed* -- there are another few assertions around polls/voting that I removed for the time being, just because I wasn't using them in bridge, but I can put them back in order to be consistent.

Reply:
Yes, please keep feature/logic parity with `reconstitute`. (: